File: js/src/reconcile.js
/**
 * Bank reconciliation: match statement lines against ledger entries.
 * Mirrors py/src/iconomics/reconcile.py.
 *
 * Only `exact` is presented as settled. Everything else is a proposal for a
 * human, because a wrong match hides a real discrepancy — the exact failure a
 * reconciliation exists to catch.
 *
 * Matching is greedy but deterministic: candidates are considered in date then
 * row order, each row is consumed at most once, so the same inputs always
 * produce the same pairing.
 */

import Decimal from 'decimal.js';

import { Money } from './money.js';
import { isoDay, normalizeCounterparty } from './parsing.js';

export const EXACT = 'exact';
export const PROBABLE = 'probable';
export const POSSIBLE = 'possible';

export const MATCHED_COLUMNS = [
  'Tier',
  'Statement Row',
  'Statement Date',
  'Statement Narration',
  'Ledger Row',
  'Ledger Date',
  'Counterparty',
  'Amount',
  'Day Gap',
];
export const UNMATCHED_STATEMENT_COLUMNS = [
  'Statement Row',
  'Date',
  'Narration',
  'Amount',
  'Reason',
];
export const UNMATCHED_LEDGER_COLUMNS = [
  'Ledger Row',
  'Date',
  'Counterparty',
  'Amount',
  'Reason',
];
export const SUMMARY_COLUMNS = ['Measure', 'Count', 'Value'];

export const MONEY_FORMATS = { Amount: '0.00', Value: '0.00' };

/** Bank value dates commonly lag the invoice date by a few days. */
export const DEFAULT_WINDOW = 5;

const MS_PER_DAY = 86400000;
const zero = (currency) => new Money(new Decimal('0.00'), currency);

/** A statement line's text: its description, or its counterparty column. */
function narration(row) {
  return [row.description ?? '', row.counterparty ?? '']
    .filter((part) => part)
    .join(' ')
    .trim();
}

/** Whether the ledger counterparty is recognisable in the statement text. */
function counterpartyAppears(ledgerRow, statementRow) {
  const name = normalizeCounterparty(ledgerRow.counterparty ?? '').toLowerCase();
  if (!name) return false;
  const haystack = normalizeCounterparty(narration(statementRow)).toLowerCase();
  if (!haystack) return false;
  if (haystack.includes(name)) return true;
  // Bank narration often truncates "Алфа ООД" to "АЛФА" or appends a reference.
  const head = name.split(' ')[0];
  return head.length >= 3 && haystack.includes(head);
}

function tierOf(ledgerRow, statementRow, window) {
  if (!ledgerRow.amountNet.equals(statementRow.amountNet)) return null;

  const gap = Math.abs(Math.round((statementRow.date - ledgerRow.date) / MS_PER_DAY));
  const named = counterpartyAppears(ledgerRow, statementRow);

  if (gap === 0 && named) return [EXACT, gap];
  if (gap <= window && named) return [PROBABLE, gap];
  if (gap <= window) return [POSSIBLE, gap];
  return null;
}

/**
 * Put every row into one currency so amounts are comparable. Real March 2026
 * data still contains BGN correction entries, and an amount cannot be matched
 * across currencies. The count is reported so the run does not quietly change
 * figures.
 */
function restate(rows, currency) {
  let changed = 0;
  const restated = rows.map((row) => {
    if (row.currency === currency) return row;
    const convert = (m) => (currency === 'EUR' ? m.toEur() : m.toBgn());
    changed += 1;
    return {
      ...row,
      amountNet: convert(row.amountNet),
      vatAmount: row.vatAmount ? convert(row.vatAmount) : null,
      currency,
    };
  });
  return { rows: restated, changed };
}

const byDateThenRow = (a, b) => a.date - b.date || a.sourceRow - b.sourceRow;

export function reconcile(statement, ledger, window = DEFAULT_WINDOW, currency = 'EUR') {
  const restatedStatement = restate(statement.rows, currency);
  const restatedLedger = restate(ledger.rows, currency);
  const restated = restatedStatement.changed + restatedLedger.changed;

  const statementRows = [...restatedStatement.rows].sort(byDateThenRow);
  const ledgerRows = [...restatedLedger.rows].sort(byDateThenRow);

  const usedLedger = new Set();
  const matched = [];

  // Best tier first, so a clean pairing is never stolen by a weaker candidate.
  for (const tier of [EXACT, PROBABLE, POSSIBLE]) {
    for (const statementRow of statementRows) {
      if (matched.some((m) => m.statement.sourceRow === statementRow.sourceRow)) continue;
      for (const ledgerRow of ledgerRows) {
        if (usedLedger.has(ledgerRow.sourceRow)) continue;
        const verdict = tierOf(ledgerRow, statementRow, window);
        if (verdict === null || verdict[0] !== tier) continue;
        matched.push({
          tier,
          statement: statementRow,
          ledger: ledgerRow,
          dayGap: verdict[1],
        });
        usedLedger.add(ledgerRow.sourceRow);
        break;
      }
    }
  }

  const matchedStatementRows = new Set(matched.map((m) => m.statement.sourceRow));
  const unmatchedStatement = statementRows.filter(
    (row) => !matchedStatementRows.has(row.sourceRow),
  );
  const unmatchedLedger = ledgerRows.filter((row) => !usedLedger.has(row.sourceRow));

  matched.sort(
    (a, b) => a.statement.date - b.statement.date || a.statement.sourceRow - b.statement.sourceRow,
  );

  return {
    matched,
    unmatchedStatement,
    unmatchedLedger,
    currency,
    window,
    restated,
    get confirmed() {
      return this.matched.filter((m) => m.tier === EXACT);
    },
    get proposed() {
      return this.matched.filter((m) => m.tier !== EXACT);
    },
  };
}

function totalOf(rows, currency) {
  let total = zero(currency);
  for (const row of rows) total = total.add(row.amountNet);
  return total;
}

export function toSheets(result) {
  const matchRows = (matches) =>
    matches.map((m) => [
      m.tier,
      m.statement.sourceRow,
      isoDay(m.statement.date),
      narration(m.statement),
      m.ledger.sourceRow,
      isoDay(m.ledger.date),
      m.ledger.counterparty,
      m.ledger.amountNet,
      m.dayGap,
    ]);

  const confirmed = result.confirmed;
  const proposed = result.proposed;

  const summaryRows = [
    [
      'Statement lines matched (confirmed)',
      confirmed.length,
      totalOf(
        confirmed.map((m) => m.statement),
        result.currency,
      ),
    ],
    [
      'Statement lines proposed (review)',
      proposed.length,
      totalOf(
        proposed.map((m) => m.statement),
        result.currency,
      ),
    ],
    [
      'Statement lines unmatched',
      result.unmatchedStatement.length,
      totalOf(result.unmatchedStatement, result.currency),
    ],
    [
      'Ledger rows unmatched',
      result.unmatchedLedger.length,
      totalOf(result.unmatchedLedger, result.currency),
    ],
    ['Date window used (days)', result.window, ''],
    [`Rows restated to ${result.currency}`, result.restated, ''],
  ];

  return {
    Matched: {
      columns: MATCHED_COLUMNS,
      rows: matchRows(confirmed),
      numberFormats: MONEY_FORMATS,
    },
    Proposed: {
      columns: MATCHED_COLUMNS,
      rows: matchRows(proposed),
      numberFormats: MONEY_FORMATS,
    },
    'Unmatched Statement': {
      columns: UNMATCHED_STATEMENT_COLUMNS,
      rows: result.unmatchedStatement.map((row) => [
        row.sourceRow,
        isoDay(row.date),
        narration(row),
        row.amountNet,
        'no ledger entry with this amount in the date window',
      ]),
      numberFormats: MONEY_FORMATS,
    },
    'Unmatched Ledger': {
      columns: UNMATCHED_LEDGER_COLUMNS,
      rows: result.unmatchedLedger.map((row) => [
        row.sourceRow,
        isoDay(row.date),
        row.counterparty,
        row.amountNet,
        'not seen on the bank statement',
      ]),
      numberFormats: MONEY_FORMATS,
    },
    Summary: {
      columns: SUMMARY_COLUMNS,
      rows: summaryRows,
      numberFormats: { Value: '0.00' },
    },
  };
}

File: py/src/iconomics/reconcile.py
"""Bank reconciliation: match statement lines against ledger entries.

Matching is tiered, and the tier is reported rather than hidden:

  exact     same amount, same date, and the counterparty appears in the
            statement narration
  probable  same amount within tolerance, date within a window, counterparty
            recognisable
  possible  amount matches, nothing else corroborates

Only ``exact`` is presented as settled. Everything else is a proposal for a human
to confirm, because a wrong match hides a real discrepancy — which is the exact
failure a reconciliation exists to catch.

Matching is greedy but deterministic: candidates are considered in date then row
order, and each ledger row and statement line is consumed at most once, so the
same inputs always produce the same pairing.
"""

from dataclasses import dataclass, replace
from decimal import Decimal

from iconomics.money import Money
from iconomics.parsing import normalize_counterparty
from iconomics.workbook import Ledger, Row, Sheet

EXACT = "exact"
PROBABLE = "probable"
POSSIBLE = "possible"

TIER_LABELS = {
    EXACT: "exact — amount, date and counterparty all agree",
    PROBABLE: "probable — amount agrees, date close, counterparty recognisable",
    POSSIBLE: "possible — amount agrees only",
}

MATCHED_COLUMNS = [
    "Tier",
    "Statement Row",
    "Statement Date",
    "Statement Narration",
    "Ledger Row",
    "Ledger Date",
    "Counterparty",
    "Amount",
    "Day Gap",
]
UNMATCHED_STATEMENT_COLUMNS = [
    "Statement Row",
    "Date",
    "Narration",
    "Amount",
    "Reason",
]
UNMATCHED_LEDGER_COLUMNS = ["Ledger Row", "Date", "Counterparty", "Amount", "Reason"]
SUMMARY_COLUMNS = ["Measure", "Count", "Value"]

MONEY_FORMATS = {"Amount": "0.00", "Value": "0.00"}

#: Default date window, in days, for a probable match. Bank value dates commonly
#: lag the invoice date by a few days.
DEFAULT_WINDOW = 5


@dataclass(frozen=True)
class Match:
    tier: str
    statement: Row
    ledger: Row
    day_gap: int


@dataclass(frozen=True)
class Reconciliation:
    matched: list[Match]
    unmatched_statement: list[Row]
    unmatched_ledger: list[Row]
    currency: str
    window: int
    restated: int = 0

    @property
    def confirmed(self) -> list[Match]:
        return [m for m in self.matched if m.tier == EXACT]

    @property
    def proposed(self) -> list[Match]:
        return [m for m in self.matched if m.tier != EXACT]


def _narration(row: Row) -> str:
    """A statement line's text: its description, or its counterparty column."""
    parts = [row.description or "", row.counterparty or ""]
    return " ".join(part for part in parts if part).strip()


def _counterparty_appears(ledger_row: Row, statement_row: Row) -> bool:
    """Whether the ledger counterparty is recognisable in the statement text."""
    name = normalize_counterparty(ledger_row.counterparty or "").casefold()
    if not name:
        return False
    haystack = normalize_counterparty(_narration(statement_row)).casefold()
    if not haystack:
        return False
    if name in haystack:
        return True
    # Fall back to the distinctive first word — bank narration often truncates
    # "Алфа ООД" to "АЛФА" or appends a reference.
    head = name.split(" ")[0]
    return len(head) >= 3 and head in haystack


def _tier(ledger_row: Row, statement_row: Row, window: int) -> tuple[str, int] | None:
    if ledger_row.amount_net != statement_row.amount_net:
        return None

    gap = abs((statement_row.date - ledger_row.date).days)
    named = _counterparty_appears(ledger_row, statement_row)

    if gap == 0 and named:
        return EXACT, gap
    if gap <= window and named:
        return PROBABLE, gap
    if gap <= window:
        return POSSIBLE, gap
    return None


def _restate(rows: list[Row], currency: str) -> tuple[list[Row], int]:
    """Put every row into one currency so amounts are comparable.

    Real March 2026 data still contains BGN correction entries, and an amount
    cannot be matched across currencies. Conversion is from the originally
    recorded amount at the fixed rate, and the count is reported so the run does
    not quietly change figures.
    """
    restated: list[Row] = []
    changed = 0
    for row in rows:
        if row.currency == currency:
            restated.append(row)
            continue
        convert = (lambda m: m.to_eur()) if currency == "EUR" else (lambda m: m.to_bgn())
        restated.append(
            replace(
                row,
                amount_net=convert(row.amount_net),
                vat_amount=convert(row.vat_amount) if row.vat_amount else None,
                currency=currency,
            )
        )
        changed += 1
    return restated, changed


def reconcile(
    statement: Ledger,
    ledger: Ledger,
    window: int = DEFAULT_WINDOW,
    currency: str = "EUR",
) -> Reconciliation:
    """Pair statement lines with ledger rows, best tier first."""
    statement_rows, restated_statement = _restate(statement.rows, currency)
    ledger_rows, restated_ledger = _restate(ledger.rows, currency)
    restated = restated_statement + restated_ledger

    statement_rows = sorted(statement_rows, key=lambda r: (r.date, r.source_row))
    ledger_rows = sorted(ledger_rows, key=lambda r: (r.date, r.source_row))

    used_ledger: set[int] = set()
    matched: list[Match] = []

    # Best tier first, so a clean pairing is never stolen by a weaker candidate.
    for tier in (EXACT, PROBABLE, POSSIBLE):
        for statement_row in statement_rows:
            if any(m.statement.source_row == statement_row.source_row for m in matched):
                continue
            for ledger_row in ledger_rows:
                if ledger_row.source_row in used_ledger:
                    continue
                verdict = _tier(ledger_row, statement_row, window)
                if verdict is None or verdict[0] != tier:
                    continue
                matched.append(
                    Match(
                        tier=tier,
                        statement=statement_row,
                        ledger=ledger_row,
                        day_gap=verdict[1],
                    )
                )
                used_ledger.add(ledger_row.source_row)
                break

    matched_statement_rows = {m.statement.source_row for m in matched}
    unmatched_statement = [
        row for row in statement_rows if row.source_row not in matched_statement_rows
    ]
    unmatched_ledger = [row for row in ledger_rows if row.source_row not in used_ledger]

    matched.sort(key=lambda m: (m.statement.date, m.statement.source_row))
    return Reconciliation(
        matched=matched,
        unmatched_statement=unmatched_statement,
        unmatched_ledger=unmatched_ledger,
        currency=currency,
        window=window,
        restated=restated,
    )


def _total(rows: list[Row], currency: str) -> Money:
    total = Money(Decimal("0.00"), currency)
    for row in rows:
        total = total + row.amount_net
    return total


def to_sheets(result: Reconciliation) -> dict[str, Sheet]:
    """Render the reconciliation. Confirmed and proposed are separate sheets."""

    def match_rows(matches: list[Match]) -> list[list[object]]:
        return [
            [
                m.tier,
                m.statement.source_row,
                m.statement.date.isoformat(),
                _narration(m.statement),
                m.ledger.source_row,
                m.ledger.date.isoformat(),
                m.ledger.counterparty,
                m.ledger.amount_net,
                m.day_gap,
            ]
            for m in matches
        ]

    confirmed = result.confirmed
    proposed = result.proposed

    summary_rows = [
        ["Statement lines matched (confirmed)", len(confirmed), _total([m.statement for m in confirmed], result.currency)],
        ["Statement lines proposed (review)", len(proposed), _total([m.statement for m in proposed], result.currency)],
        ["Statement lines unmatched", len(result.unmatched_statement), _total(result.unmatched_statement, result.currency)],
        ["Ledger rows unmatched", len(result.unmatched_ledger), _total(result.unmatched_ledger, result.currency)],
        ["Date window used (days)", result.window, ""],
        ["Rows restated to " + result.currency, result.restated, ""],
    ]

    return {
        "Matched": Sheet(
            columns=MATCHED_COLUMNS,
            rows=match_rows(confirmed),
            number_formats=MONEY_FORMATS,
        ),
        "Proposed": Sheet(
            columns=MATCHED_COLUMNS,
            rows=match_rows(proposed),
            number_formats=MONEY_FORMATS,
        ),
        "Unmatched Statement": Sheet(
            columns=UNMATCHED_STATEMENT_COLUMNS,
            rows=[
                [
                    row.source_row,
                    row.date.isoformat(),
                    _narration(row),
                    row.amount_net,
                    "no ledger entry with this amount in the date window",
                ]
                for row in result.unmatched_statement
            ],
            number_formats=MONEY_FORMATS,
        ),
        "Unmatched Ledger": Sheet(
            columns=UNMATCHED_LEDGER_COLUMNS,
            rows=[
                [
                    row.source_row,
                    row.date.isoformat(),
                    row.counterparty,
                    row.amount_net,
                    "not seen on the bank statement",
                ]
                for row in result.unmatched_ledger
            ],
            number_formats=MONEY_FORMATS,
        ),
        "Summary": Sheet(
            columns=SUMMARY_COLUMNS,
            rows=summary_rows,
            number_formats={"Value": "0.00"},
        ),
    }
